### PR TITLE
source-sparkpost contribution from bishalbera

### LINKED ISSUE #46062: [connector-request] SparkPost
This is a **suggestion** to create a connector for **SparkPost** using the UI Connector Builder.
The website for the connector is https://www.sparkpost.com.

If you want to try, claim this issue and start working on it following the steps below.

### Steps:
1. Comment in the issue and wait to be assigned to start working on it.
2. Map the API endpoints (get approval to move to 3 step)
3. Describe steps to get credentials.
4. Create the connector using the UI Builder

> [!IMPORTANT]
> Airbyte doesn't have sandbox credentials for this connector. You must have access/credentials to the service provider to create the connector.

> [!CAUTION]
> This is a **suggestion** and there may be cases where creating the connector using the Builder won't be possible. To minimize the risk of investing a lot of time directly in creating the connector, we strongly recommend to follow steps above.

### PR DESCRIPTION
**Pull request by bishalbera**
## What

This PR adds a new connector for source SparkPost (source-sparkpost).

Contributor-provided connector description:

The SparkPost connector for Airbyte enables seamless integration with SparkPost’s email delivery service, allowing users to automatically sync email performance data, including delivery, open, and click metrics, into their data warehouses.
## Details

<details><summary>Expand for list of spec properties and streams</summary>

### Configuration

| Input | Type | Description | Default Value |
|-------|------|-------------|---------------|
| `api_key` | `string` | API Key.  |  |
| `start_date` | `string` | Start Date.  |  |
| `api_prefix` | `string` | API Endpoint Prefix.  | api |

### Streams
| Stream Name | Primary Key | Pagination | Supports Full Sync | Supports Incremental |
|-------------|-------------|------------|---------------------|----------------------|
| message_events | event_id | DefaultPaginator | ✅ |  ❌  |
| sending_domains | domain | No pagination | ✅ |  ❌  |
| ab_test | id | No pagination | ✅ |  ❌  |
| templates | id | No pagination | ✅ |  ❌  |
| recipients | id | No pagination | ✅ |  ❌  |
| subaccounts | id | DefaultPaginator | ✅ |  ❌  |
| snippets | id | No pagination | ✅ |  ❌  |

</details>

## Reviewer checklist
- [ ] Run `/format-fix` to resolve any formatting errors
- [ ] Help the contributor upload an icon that follows our icon design guidelines
- [ ] Click into the CI workflows that wait for a maintainer to run them, which should trigger CI runs

### CODE REVIEW

**Comment by bishalbera** [comment]
@marcosmarxm There is rate limiting for this but I couldn't find a way to implement it. So if you guide me or give me some docs ref maybe I can implement this.
Screenshot 2024-10-10 at 8 57 24 AM

File: airbyte-integrations/connectors/source-sparkpost/manifest.yaml [hunk 1]
```diff
@@ -0,0 +1,1007 @@
...
+            type: DpathExtractor
+            field_path:
+              - results
+      schema_loader:
+        type: InlineSchemaLoader
+        schema:
+          $ref: "#/schemas/snippets"
+  base_requester:
+    type: HttpRequester
+    url_base: https://api.sparkpost.com
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-sparkpost/manifest.yaml:302]: Sparkpost have two endpoints one for global and other for EU region, check [here](https://developers.sparkpost.com/api/#header-endpoints)

File: docs/integrations/sources/sparkpost.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+| ab_test | id | No pagination | ✅ |  ❌  |
+| templates | id | No pagination | ✅ |  ❌  |
+| recipients | id | No pagination | ✅ |  ❌  |
+| metrics_summary |  | No pagination | ✅ |  ❌  |
+| subaccounts | id | DefaultPaginator | ✅ |  ❌  |
+| engagement_details |  | No pagination | ✅ |  ❌  |
+| rejection_reason_metrics |  | DefaultPaginator | ✅ |  ❌  |
+| metrics |  | No pagination | ✅ |  ❌  |
+| metrics_by_template | template_id | DefaultPaginator | ✅ |  ❌  |
+| snippets | id | No pagination | ✅ |  ❌  |
```
Comment (marcosmarxm) [inline docs/integrations/sources/sparkpost.md:33]: I'd suggest to try manually checking if pagination works. If not most endpoints supports from-to parameters where you can get incremental + read all data in batches.

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
Initial review.

File: airbyte-integrations/connectors/source-sparkpost/manifest.yaml (hunk 1 above)
Comment (bishalbera) [reply airbyte-integrations/connectors/source-sparkpost/manifest.yaml:302]: @marcosmarxm they both work independently with diff api keys. If one signedup for global they has to use it.

File: docs/integrations/sources/sparkpost.md (hunk 2 above)
Comment (bishalbera) [reply docs/integrations/sources/sparkpost.md:33]: @marcosmarxm I think incremental sync can be done as I've seen from and to parameters in some endpoints . will try that

Comment (bishalbera) [reply docs/integrations/sources/sparkpost.md]: implemented constant backoff strategy of 5 secs for rate limiting and added pagination to some endpoints. @marcosmarxm

File: airbyte-integrations/connectors/source-sparkpost/manifest.yaml (hunk 1 above)
Comment (marcosmarxm) [reply airbyte-integrations/connectors/source-sparkpost/manifest.yaml:408]: What I'm saying your connector must have a `region` parameter. You don't need to test EU but prepare it to work.

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
`recipients`: ✅ no pagination and limits
`sending_domain`: ✅  no pagination and limits
`ab_test`: ✅ no pagination and limits

`templates`: there is a limitation in the API doesn't support pagination but returns 50k. Please include this information in the docs.
> Returns an array containing up to your first 50,000 templates.

`search_message_events`: 
- rename to `message_events`
- Why are you requesting `templates` and `recipients`? It won't return all messages without this?

**domain config parameter** it isn't used anywhere please remove and implement `region` used in base url.

The pagination for this API is really not the best. Can you start a discussion in Slack asking what is the best strategy to retrieve the cursor when the result is something like:
```json
    {
      "results": [
        ...
      ],
      "total_count": 20000,
      "links": {
        "next": "/api/v1/events/message?events=delivery&per_page=1000&cursor=WycyMDE4LTExLTA1VDIyOjQ1OjM5LjAwMFonLCAnc3BjLTM4MTQ1MjY3MjMyNTA2NTEwJ10="
      }
    }
```

**Comment by bishalbera** [comment]
> `recipients`: ✅ no pagination and limits `sending_domain`: ✅ no pagination and limits `ab_test`: ✅ no pagination and limits
> 
> `templates`: there is a limitation in the API doesn't support pagination but returns 50k. Please include this information in the docs.
> 
> > Returns an array containing up to your first 50,000 templates.
> 
> `search_message_events`:
> 
> * rename to `message_events`
> * Why are you requesting `templates` and `recipients`? It won't return all messages without this?
> 
> **domain config parameter** it isn't used anywhere please remove and implement `region` used in base url.
> 
> The pagination for this API is really not the best. Can you start a discussion in Slack asking what is the best strategy to retrieve the cursor when the result is something like:
> 
> ```json
>     {
>       "results": [
>         ...
>       ],
>       "total_count": 20000,
>       "links": {
>         "next": "/api/v1/events/message?events=delivery&per_page=1000&cursor=WycyMDE4LTExLTA1VDIyOjQ1OjM5LjAwMFonLCAnc3BjLTM4MTQ1MjY3MjMyNTA2NTEwJ10="
>       }
>     }
> ```

I asked this in slack. I thought that `templates` and `recipients` are required parameters as It was used in the sample req but it seems they are not. So I removed it. @marcosmarxm

**Comment by marcosmarxm** [comment]
Hello 👋 Due to the numerous contributions happening, I decided to create this [view](https://github.com/orgs/airbytehq/projects/104/views/3) to help you know when I plan to review your contribution. If your connector has some limited-time credential feel free to reach me out.

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
I made a couple changes in your submissions directly in the builder. Today the endpoint `message_events` is missing data. This stream requires more than one records to validate pagination is working properly. Please let me know when this is done to review the contribution again.

**Review by marcosmarxm — APPROVED** [review]

**Pull request merged**